iwyu_output.cc

```diff
@@ -1737,8 +1737,29 @@ static string GetWarningMsg(const OneUse& use) {
 int IwyuFileInfo::EmitWarningMessages(const vector<OneUse>& uses) {
   set<pair<int, string>> iwyu_warnings;   // line-number, warning-msg.
   for (const OneUse& use : uses) {
-    if (use.is_iwyu_violation())
-      iwyu_warnings.insert(make_pair(use.UseLinenum(), GetWarningMsg(use)));
+    if (use.is_iwyu_violation()) {
+      const int linenum = use.UseLinenum();
+      if (!use.is_full_use()) {
+        // Avoid fwd-decl warnings nearby complete-use warnings to make output
+        // cleaner.
+        auto full_use_in_line = [linenum,
+                                 decl = use.decl()](const OneUse& other_use) {
+          if (!other_use.is_full_use() || other_use.UseLinenum() != linenum)
+            return false;
+          if (other_use.decl() == decl)
+            return true;
+          // For class template implicit instantiations, fwd-decl uses refer to
+          // ClassTemplateDecl whereas full uses refer to their definition
+          // nodes. Partial specs aren't considered for simplicity.
+          return isa<ClassTemplateDecl>(decl) &&
+                 GetTagDefinition(decl) == other_use.decl();
+        };
+        if (std::any_of(uses.begin(), uses.end(), full_use_in_line))
```

Comment:
The nested loop here is a little annoying, I guess a file with many uses (like badinc.cc) might suffer.

Would it be possible to do something like this instead:
```diff
diff --git a/iwyu_output.cc b/iwyu_output.cc
index 99958ef..dd15ade 100644
--- a/iwyu_output.cc
+++ b/iwyu_output.cc
@@ -1736,29 +1736,34 @@ static string GetWarningMsg(const OneUse& use) {
 
 int IwyuFileInfo::EmitWarningMessages(const vector<OneUse>& uses) {
   set<pair<int, string>> iwyu_warnings;   // line-number, warning-msg.
+
+  // Report all full uses, and build an index of their (line,decl).
+  set<pair<int, const NamedDecl*>> full_use_index;
   for (const OneUse& use : uses) {
-    if (use.is_iwyu_violation()) {
-      const int linenum = use.UseLinenum();
-      if (!use.is_full_use()) {
-        // Avoid fwd-decl warnings nearby complete-use warnings to make output
-        // cleaner.
-        auto full_use_in_line = [linenum,
-                                 decl = use.decl()](const OneUse& other_use) {
-          if (!other_use.is_full_use() || other_use.UseLinenum() != linenum)
-            return false;
-          if (other_use.decl() == decl)
-            return true;
-          // For class template implicit instantiations, fwd-decl uses refer to
-          // ClassTemplateDecl whereas full uses refer to their definition
-          // nodes. Partial specs aren't considered for simplicity.
-          return isa<ClassTemplateDecl>(decl) &&
-                 GetTagDefinition(decl) == other_use.decl();
-        };
-        if (std::any_of(uses.begin(), uses.end(), full_use_in_line))
-          continue;
-      }
+    if (!use.is_iwyu_violation() || !use.is_full_use())
+      continue;
+
+    int line = use.UseLinenum();
+    iwyu_warnings.insert(std::make_pair(line, GetWarningMsg(use)));
+    full_use_index.insert(std::make_pair(line, use.decl()));
+  }
+
+  // Report all forward-declaration, except for (line,decl) already covered by a
+  // full use.
+  for (const OneUse& use : uses) {
+    if (!use.is_iwyu_violation() || use.is_full_use())
+      continue;
+
+    // For class template implicit instantiations, fwd-decl uses refer to
+    // ClassTemplateDecl whereas full uses refer to their definition nodes.
+    // Partial specs aren't considered for simplicity.
+    const NamedDecl* decl = use.decl();
+    if (isa<ClassTemplateDecl>(decl))
+      decl = GetTagDefinition(decl);
 
-      iwyu_warnings.insert(make_pair(linenum, GetWarningMsg(use)));
+    int line = use.UseLinenum();
+    if (!full_use_index.count(std::make_pair(line, decl))) {
+      iwyu_warnings.insert(std::make_pair(line, GetWarningMsg(use)));
     }
   }
   // Nice that set<> automatically sorts things for us!
```

I played with it a little, and it appears to save ~1 second for the test suite on average.

Reply:
Nice optimization, thank you!